### PR TITLE
Merge Python 2 and 3 shutil

### PR DESCRIPTION
**Pull request by srittau**

### CODE REVIEW

File: stdlib/2and3/shutil.pyi [hunk 1]
```diff
@@ -1,34 +1,51 @@
...
 from typing import (
     List, Iterable, Callable, Any, Tuple, Sequence, NamedTuple, IO,
-    AnyStr, Optional, Union
+    AnyStr, Optional, Union, Set, TypeVar, overload, Type
 )
 
 if sys.version_info >= (3, 6):
     _Path = Union[str, os.PathLike[str]]
+    _AnyStr = str
+    _AnyPath = TypeVar("_AnyPath", str, os.PathLike[str])
```
Comment (JelleZijlstra) [inline stdlib/2and3/shutil.pyi]: Reminder to self: I'm worried about these conditionally defined typevars. Do they make sense for all use cases?

Comment (srittau) [reply stdlib/2and3/shutil.pyi]: Side note: Unfortunately, in Python 3 shutil is a bit of a mess, type-wise. As is alluded to in the stub, bytes is often does not work as a path and whether the returned path is a `str` or a `Path` is dependent on the phase of the moon. I am seriously contemplating to propose patches for the stdlib to fix this for 3.8, but of course that wouldn't help for the 3.6 and 3.7 stubs.

File: stdlib/2and3/shutil.pyi [hunk 2]
```diff
@@ -41,36 +58,41 @@ else:
...
-def rmtree(path: _Path, ignore_errors: bool = ...,
-           onerror: Callable[[Any, Any, Any], None] = ...) -> None: ...
+    _IgnoreFn = Union[None, Callable[[AnyStr, List[AnyStr]], Iterable[AnyStr]]]
+    def copytree(src: AnyStr, dst: AnyStr, symlinks: bool = ...,
+                 ignore: _IgnoreFn = ...) -> _PathReturn: ...
+
+if sys.version_info >= (3,):
+    @overload
+    def rmtree(path: bytes, ignore_errors: bool = ...,
+               onerror: Callable[[Any, str, Any], None] = ...) -> None: ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/shutil.pyi]: The return type of the callback should be `Any` for flexibility (for example, a user could pass in a lambda that accidentally returns some value).

File: stdlib/2and3/shutil.pyi [hunk 3]
```diff
@@ -80,39 +102,25 @@ if sys.version_info >= (3, 3):
...
                  dry_run: bool = ..., owner: str = ..., group: str = ...,
-                 logger: Any = ...) -> str: ...
+                 logger: Any = ...) -> _AnyStr: ...
 def get_archive_formats() -> List[Tuple[str, str]]: ...
 
-# TODO function is a callback that receives keyword arguments; should make it not use Any
-# once we have support for callable types with keyword args
-def register_archive_format(name: str, function: Any,
+def register_archive_format(name: str, function: Callable[..., Any],
                             extra_args: Sequence[Tuple[str, Any]] = ...,
```
Comment (JelleZijlstra) [inline stdlib/2and3/shutil.pyi]: extra_args is Optional.

Also, the implementation actually allows only tuples and lists, not all sequences (there's an isinstance check). And within the sequence, the elements can be either tuples or lists. The first change would be a pain to support properly due to invariance, but we can accommodate the second one.

In summary, I think this should be `Optional[Sequence[Union[Tuple[str, Any], List[Any]]]`.

File: stdlib/2and3/shutil.pyi [hunk 4]
```diff
@@ -80,39 +102,25 @@ if sys.version_info >= (3, 3):
...
-    class ReadError(OSError): ...
-else:
-    class Error(EnvironmentError): ...
-    class SpecialFileError(EnvironmentError): ...
-    class ExecError(EnvironmentError): ...
-    class ReadError(EnvironmentError): ...
-class RegistryError(Exception): ...
-
-def make_archive(base_name: str, format: str, root_dir: _Path = ...,
+def make_archive(base_name: _AnyStr, format: str, root_dir: _Path = ...,
```
Comment (JelleZijlstra) [inline stdlib/2and3/shutil.pyi]: `root_dir`, `base_dir`, `owner`, `group`, and `logger` are Optional.

File: stdlib/2and3/shutil.pyi [hunk 5]
```diff
@@ -80,39 +102,25 @@ if sys.version_info >= (3, 3):
...
-def register_unpack_format(name: str, extensions: List[str], function: Any,
-                           extra_args: Sequence[Tuple[str, Any]] = ...,
-                           description: str = ...) -> None: ...
-def unregister_unpack_format(name: str) -> None: ...
-def get_unpack_formats() -> List[Tuple[str, List[str], str]]: ...
 
-if sys.version_info >= (3, 3):
+if sys.version_info >= (3,):
+    # Should be _Path once http://bugs.python.org/issue30218 is fixed
+    def unpack_archive(filename: str, extract_dir: _Path = ...,
```
Comment (JelleZijlstra) [inline stdlib/2and3/shutil.pyi]: extract_dir and format are Optional.

File: stdlib/2and3/shutil.pyi [hunk 6]
```diff
@@ -41,36 +58,41 @@ else:
...
+if sys.version_info >= (3,):
+    @overload
+    def rmtree(path: bytes, ignore_errors: bool = ...,
+               onerror: Callable[[Any, str, Any], None] = ...) -> None: ...
+    @overload
+    def rmtree(path: _AnyPath, ignore_errors: bool = ...,
+               onerror: Callable[[Any, _AnyPath, Any], None] = ...) -> None: ...
+else:
+    def rmtree(path: _AnyPath, ignore_errors: bool = ...,
+               onerror: Callable[[Any, _AnyPath, Any], None] = ...) -> None: ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/shutil.pyi]: onerror is Optional.

**Comment by JelleZijlstra** [comment]
I just reviewed the stubs against the master and 2.7 implementations of shutil and noticed a few minor discrepancies. Most of them are probably not new, but I think this is a good time to fix them anyway.

**Comment by srittau** [comment]
Agreed, will fix.

**Comment by JelleZijlstra** [comment]
Thanks!

**Pull request merged**